### PR TITLE
[feat] Gather experience samples

### PR DESCRIPTION
**Pull request by reciprocated**
This PR lets PPO trainer to gather all experience samples on the main rank and to do a single joint reward_fn call per each rollout.

This enables hosting a single reward model on the same machine as the main rank, when previously every process had to had access to the reward model. Also it enables deliberate micro-batching for the reward model, unlike in the case when each process tries to infer reward model (for example deployed on Triton server) with its own small chunk_size number of samples usually bottle-necking whole training.

Main goal here is to give up the current dependency on the Triton server and to enable simple and self-contained 7+1(RM) or 15+1(RM) setups (can finally move to cw)

https://wandb.ai/sorry/trlx/reports/Gather-experience-samples-305---VmlldzozNTMxMzc3

Side-note: every reference remains the same except for sentiments, after some debugging I've noticed that even doing multiple redundant passes of sentiment pipeline on the same data apparently changes rng or otherwise slightly influences the run, as in doing:
```python
scores = torch.tensor(
    self.reward_fn(
        samples=str_samples,
        prompts=str_prompts,
        outputs=str_outputs,
    ),
    dtype=torch.float,
).to(device)

scores = torch.tensor(
    self.reward_fn(
        samples=str_samples,
        prompts=str_prompts,
        outputs=str_outputs,
    ),
    dtype=torch.float,
).to(device)

scores = torch.tensor(
    self.reward_fn(
        samples=str_samples,
        prompts=str_prompts,
        outputs=str_outputs,
    ),
    dtype=torch.float,
).to(device)
```

### CODE REVIEW

**Comment by cat-state** [comment]
Maybe we need to call `.eval()` somewhere, if its using dropout? Or it could be nondeterminism from the kernels.

**Comment by reciprocated** [comment]
Oh I've forgot to mention that scores from the sentiment pipeline remain the same, as doing so passes:
```python
scores = torch.tensor(
    self.reward_fn(
        samples=str_samples,
        prompts=str_prompts,
        outputs=str_outputs,
    ),
    dtype=torch.float,
).to(device)

scores2 = torch.tensor(
    self.reward_fn(
        samples=str_samples,
        prompts=str_prompts,
        outputs=str_outputs,
    ),
    dtype=torch.float,
).to(device)

assert torch.all(scores == scores2)
```

However the number of calls to `reward_fn` is divided by `num_processes` as compared to the reference, so I suspect that's reason behind the difference

**Review by jon-tow — CHANGES_REQUESTED** [review]
Awesome! Glad to see you push us to the CW side hehe Left one comment if you could address before merging 🙏

**Comment by jon-tow** [comment]
> re the sidenote: Maybe we need to call .eval() somewhere, if its using dropout? Or it could be nondeterminism from the kernels.

The sentiments pipeline is in eval-mode so at least it's not any stochasticity from dropout in that distilbert RM (`sentiment_fn.model.training == False`) 🤔 

Also, the returns/values stats sort of explode for sentiments on main - it's interesting that the repeat call smooths things out.

**Comment by LouisCastricato** [comment]
This is great! Happy that we're finally merging something like this.

File: trlx/trainer/accelerate_ppo_trainer.py [hunk 1]
```diff
@@ -286,6 +286,42 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):  # noq
...
+                stats["time/exp_score"] = time() - exp_score_time
+
+                all_scores = list(all_scores.reshape(torch.distributed.get_world_size(), -1).unbind())
+            else:
+                all_scores = None
+
+            scores = torch.empty(len(samples), device=device)
+            torch.distributed.scatter(scores, all_scores)
+
             str_samples, str_prompts, str_outputs = self.decode(prompt_tensors, samples)
```
[the remark this answers is not in the document]
Comment (reciprocated) [reply trlx/trainer/accelerate_ppo_trainer.py:325]: To avoid repetetion I could scatter samples as well, but I don't think it's worth it since the second `decode` (which would be second only on the main rank) is needed just to strip `stop_sequences`, so it's basically just an allocation. Runtime in `make_experience` is still dominated by `generate` and `reward_fn`

[the remark this answers is not in the document]
Comment (jon-tow) [reply trlx/trainer/accelerate_ppo_trainer.py:325]: Yeah, that's fine! (I actually meant to approve this last night but forgot to 😅)

**Review by jon-tow — APPROVED** [review]
Thanks Max!

**Comment by reciprocated** [comment]
@jon-tow Even if the pipeline is in eval-mode, rng state is still accessed somewhere from within it, a simple test:
```python
torch.manual_seed(1000)
print(f'{torch.rand(1)=}')
print(f'{torch.rand(1)=}')
print(f'{torch.rand(1)=}')
>>> torch.rand(1)=tensor([0.3189])
>>> torch.rand(1)=tensor([0.6136])
>>> torch.rand(1)=tensor([0.4418])

torch.manual_seed(1000)
print(f'{torch.rand(1)=}')
print(f'{torch.rand(1)=}')
reward_fn(['1'])
print(f'{torch.rand(1)=}')
>>> torch.rand(1)=tensor([0.3189])
>>> torch.rand(1)=tensor([0.6136])
>>> torch.rand(1)=tensor([0.2724])
```
Since none of ranks except the main now access `reward_fn`, a difference surfaced. 

https://wandb.ai/sorry/trlx/reports/Gather-experience-samples-305--VmlldzozNTM1OTUz

Here's two additional reports: a single process run, and a run with changed `reward_fn` from the pipeline to something deterministic
```python
def reward_fn(samples, **kwargs):
    return [len(s) for s in samples]
```

**Comment by jon-tow** [comment]
@reciprocated yeah that's weird. If you call the underlying distilbert with some arbitrary inputs the RNG doesn't change  https://colab.research.google.com/drive/1FCmeWEJGl5GAhikeUXXR5VrdHOt4S6WN?usp=sharing
`pipeline` is super opaque so I can't tell where it's happening 🤔 This PR should be fine especially since the t5 summarize reference is good.

**Comment by LouisCastricato** [comment]
We should ping hugging face folks

**Comment by reciprocated** [comment]
Final double check:
https://wandb.ai/sorry/trlx/reports/Gather-experience-samples-305--VmlldzozNTQ0OTkz
